pkg/controller/master/upgradelog/common.go

```diff
@@ -0,0 +1,959 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{
+				harvesterUpgradeLogLabel: upgradeLog.Name,
+			},
+			Name:      fmt.Sprintf("%s-operator", upgradeLog.Name),
+			Namespace: managedChartNamespace,
+		},
+		Spec: mgmtv3.ManagedChartSpec{
+			Chart:            "rancher-logging",
+			ReleaseName:      fmt.Sprintf("%s-operator", upgradeLog.Name),
+			DefaultNamespace: upgradeLogNamespace,
+			RepoName:         "harvester-charts",
+			Targets: []v1alpha1.BundleTarget{
+				{
+					ClusterName: "local",
+					ClusterSelector: &metav1.LabelSelector{
+						MatchExpressions: []metav1.LabelSelectorRequirement{
+							{
+								Key:      "provisioning.cattle.io/unmanaged-system-agent",
+								Operator: metav1.LabelSelectorOpDoesNotExist,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func prepareLogging(upgradeLog *harvesterv1.UpgradeLog) *loggingv1.Logging {
+	return &loggingv1.Logging{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: map[string]string{
+				harvesterUpgradeLogLabel: upgradeLog.Name,
+			},
+			Name: fmt.Sprintf("%s-infra", upgradeLog.Name),
+		},
+		Spec: loggingv1.LoggingSpec{
+			ControlNamespace:        upgradeLog.Namespace,
+			FlowConfigCheckDisabled: true,
+			FluentbitSpec: &loggingv1.FluentbitSpec{
+				Labels: map[string]string{
+					harvesterUpgradeLogLabel:          upgradeLog.Name,
+					harvesterUpgradeLogComponentLabel: ShipperComponent,
+				},
+				Image: loggingv1.ImageSpec{
+					Repository: fluentBitImageRepo,
+					Tag:        fluentBitImageTag,
+				},
+				Tolerations: []corev1.Toleration{
+					{
+						Key:      "node-role.kubernetes.io/master",
+						Operator: corev1.TolerationOpExists,
+						Effect:   corev1.TaintEffectNoSchedule,
+					},
+				},
+			},
+			FluentdSpec: &loggingv1.FluentdSpec{
+				Labels: map[string]string{
+					harvesterUpgradeLogLabel:          upgradeLog.Name,
+					harvesterUpgradeLogComponentLabel: AggregatorComponent,
+				},
+				Image: loggingv1.ImageSpec{
+					Repository: fluentdImageRepo,
+					Tag:        fluentdImageTag,
+				},
+				ConfigReloaderImage: loggingv1.ImageSpec{
+					Repository: configReloaderImageRepo,
+					Tag:        configReloaderImageTag,
+				},
+				DisablePvc: true,
+				ExtraVolumes: []loggingv1.ExtraVolume{
+					{
+						ContainerName: "fluentd",
+						Path:          "/archive",
+						VolumeName:    "log-archive",
+						Volume: &volume.KubernetesVolume{
+							PersistentVolumeClaim: &volume.PersistentVolumeClaim{
+								PersistentVolumeSource: corev1.PersistentVolumeClaimVolumeSource{
+									ClaimName: fmt.Sprintf("%s-log-archive", upgradeLog.Name),
+									ReadOnly:  false,
+								},
+							},
+						},
+					},
+				},
+				FluentOutLogrotate: &loggingv1.FluentOutLogrotate{
+					Age:     "10",
+					Enabled: true,
+					Path:    "/fluentd/log/out",
+					Size:    "10485760",
+				},
+			},
+		},
+	}
+}
+
+func prepareClusterFlow(upgradeLog *harvesterv1.UpgradeLog) *loggingv1.ClusterFlow {
```

Comment:
could we put all of those definition into one yaml file, and use `kubectl apply -f ` to create

the indepedent yaml file will make it easier to maintain and extend

e.g.

https://github.com/harvester/harvester-installer/blob/master/pkg/config/templates/rancherd-10-harvester.yaml#L69

Reply:
@w13915984028 I'd suggest we keep the way to create these objects programmatically. These resources are rarely changed and we can save the hassle to embed or manipulate YAML files. Let's put our effort into how to work well with the logging add-on transition first.